src/3devoUpdate/Form1.cs

```diff
@@ -1,610 +1,541 @@
-                }
-                catch (Exception)
-                {
-                    return;
-                }
-
-                Array.Sort(devs);
-
-                // Loop through each device
-                foreach (string dev in devs)
-                {
-                    // See if device starts with one of the prefixes
-                    foreach (string prefix in devPrefixs)
-                    {
-                        if (dev.StartsWith("/dev/" + prefix))
-                        {
-                            cmbPort.Items.Add(dev);
-                            break;
-                        }
-                    }
-                }
-            }
-            else // Windows
-            {
-                //update_com_ports();
-            }
-        }
+    private void EnableInterface(bool enableUpload = true) {
+      Util.InvokeIfRequired(this, c => {
+        if( enableUpload )
+          btnUpload.Enabled = true;
 
-        // FlashFile is changed event
-        private void FlashFile_Changed(object sender, EventArgs e)
-        {
-            update_status_info();
-        }
+        btnFlashBrowse.Enabled = true;
+        cmbPort.Enabled = true;
+      });
+    }
+
+    private void DisableInterface() {
+      Util.InvokeIfRequired(this, c => {
+        btnUpload.Enabled = false;
+        btnFlashBrowse.Enabled = false;
+        cmbPort.Enabled = false;
+      });
+    }
 
-        // Browse for flash file
-        private void btnFlashBrowse_Click(object sender, EventArgs e)
-        {
-            if (openFileDialog1.ShowDialog() == DialogResult.OK)
-            {
-                txtFlashFile.Text = openFileDialog1.FileName;                
-            }                
+    private void UpdateInterface() {
+      // Clear status info text
+      txtStatusInfo.Clear();
+
+      cmbPort.Enabled = true;
+
+      // No devices connected/found
+      if( cmbPort.Items.Count == 0 ) {
+        btnFlashBrowse.Enabled = false;
+        txtStatusInfo.AppendText(NO_MACHINE_CONNECTED);
+      }
+      // 1 or more devices connected
+      else {
+        btnFlashBrowse.Enabled = true;
+
+        if( deviceSelected == false ) {
+          txtStatusInfo.AppendText(PORT_NOT_SELECTED);
+        } else if( flashFile.Length == 0 ) {
+          txtStatusInfo.AppendText(FILE_NOT_SELECTED);
+        } else {
+          txtStatusInfo.AppendText(READY_FOR_UPLOADING);
         }
+      }
+
+      // User warning when the machine is detected in bootloader mode.
+      if (bootloaderDevice == true) {
+        txtStatusInfo.AppendText("\r\nGeneric ST device selected;\n");
+        txtStatusInfo.AppendText("Carefully select the correct binary file for your machine, since we cannot determine which 3devo device is connected!\n");
+        txtStatusInfo.AppendText("Uploading the wrong firmware may cause hardware damage or the machine might stop responding.\n");

```

Comment:
Maybe add "Only use this when recovering from a failed firmware update using the BOOT pin"?